### PR TITLE
Unstructured Mesh Sampling Updates

### PR DESCRIPTION
**Pull request by pshriwise**
This PR is an extension of @NybergWISC's fine work in #2151. It includes some minor code changes, documentation, and a rework of the tests to use the `openmc.lib.sample_external_source` method (sorry I didn't think of that as an option sooner, Matthew). I've also rebased it onto the current version of `develop`, so I thought it would be easier to create a new PR rather than PR into the branch from #2151.

If #2329 goes in before this PR, we'll update the CDF generated in the `MeshSpatial` class to use that work.

As follow-on PR's to this I'd like to:

  - add support for sampling of other mesh types `Regular`, `Cylindrical`, etc.
  - add a `MeshSource` class that will handle user-defined angle/energy/time distributions that can be set either for the whole mesh or element-by-element

### CODE REVIEW

File: src/distribution_spatial.cpp [hunk 1]
```diff
@@ -180,6 +182,71 @@ Position SphericalIndependent::sample(uint64_t* seed) const
...
+  mesh_CDF_.back() = 1.0;
+}
+
+Position MeshSpatial::sample(uint64_t* seed) const
+{
+  // Create random variable for sampling element from mesh
+  double eta = prn(seed);
+  // Sample over the CDF defined in initialization above
+  int32_t elem_idx = lower_bound_index(mesh_CDF_.begin(), mesh_CDF_.end(), eta);
+  return mesh_ptr_->sample(seed, elem_idx);
```
Comment (NybergWISC) [inline src/distribution_spatial.cpp:247]: Just to document my challenges with the code; if one needs to use the mesh index sampled here to dependently sample over energy/direction then they must either use a costly function to find the mesh element id from position when sampling energy or return a second variable (elem_idx most likely) here. Returning a second variable here does not fit the formatting of the IndependentSource class so a small refactor may be required in the future to move this code into a new "DependentSource" class.

File: openmc/stats/multivariate.py [hunk 2]
```diff
@@ -616,6 +621,128 @@ def from_xml_element(cls, elem):
...
+    Parameters
+    ----------
+    mesh : openmc.MeshBase
+        The mesh instance used for sampling, mesh is written into settings.xml,
+        mesh.id is written into the source distribution
+    strengths : Iterable of Real, optional
+        A list of values which represent the weights of each element. If no source
+        strengths are specified, they will be equal for all mesh elements.
+    volume_normalized : bool, optional
+        Whether or not the strengths will be normalized by volume. Default is True.
```
Comment (NybergWISC) [inline openmc/stats/multivariate.py:641]: For clarity as in line 675:

```suggestion
        Whether or not the strengths will be normalized (multiplied) by volume. Default is True.
```

Comment (pshriwise) [reply openmc/stats/multivariate.py:641]: Hrmmm yeah I think "multiplied" is simply more correct here. Updating now.

File: src/distribution_spatial.cpp (hunk 1 above)
Comment (pshriwise) [reply src/distribution_spatial.cpp:247]: Spot on. I found that too and have implemented such a method. Thanks for noting this here!

**Comment by pshriwise** [comment]
Ended up deviating from the suggestions a bit, but added you as co-author on the latest commit @NybergWISC. Thanks for looking this over!

**Comment by paulromano** [comment]
@pshriwise @NybergWISC thanks for the great work in this PR and the successor PR #2151. Just starting a review now and should have full comments shortly.

@pshriwise One thing I noticed right away is that the formatting of some of the C++ source files seems off. Do you have the clang-format [post-commit hook installed](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting)?

File: tests/regression_tests/unstructured_mesh/source_sampling/test.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    water_mat.add_nuclide("H1", 2.0)
+    water_mat.add_nuclide("O16", 1.0)
+    water_mat.set_density("atom/b-cm", 0.07416)
+    materials.append(water_mat)
+
+    ### Geometry ###
+    # create a regular mesh that matches the superimposed mesh
+    regular_mesh = openmc.RegularMesh(mesh_id=10)
+    regular_mesh.lower_left = (-10, -10, -10)
+    regular_mesh.dimension = (10, 10, 10)
```
Comment (paulromano) [inline tests/regression_tests/unstructured_mesh/source_sampling/test.py:34]: This results in 1000 elements/cells -- maybe let's make this a bit smaller

File: tests/regression_tests/unstructured_mesh/source_sampling/test.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    regular_mesh.width = (2, 2, 2)
+
+    root_cell, _ = regular_mesh.build_cells()
+
+    geometry = openmc.Geometry(root=[root_cell])
```
Comment (paulromano) [inline tests/regression_tests/unstructured_mesh/source_sampling/test.py:39]:
```suggestion
    geometry = openmc.Geometry([root_cell])
```

File: tests/regression_tests/unstructured_mesh/source_sampling/test.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+    # set boundary conditions of the root cell
+    for surface in root_cell.region.get_surfaces().values():
+        surface.boundary_type = 'vacuum'
```
Comment (paulromano) [inline tests/regression_tests/unstructured_mesh/source_sampling/test.py:43]: Looks like `build_cells` above has a `bc` argument that could be set to 'vacuum', thus avoiding the need for this

File: tests/regression_tests/unstructured_mesh/source_sampling/test.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    # subtract one to account for root cell
+    n_cells = len(geometry.get_all_cells()) - 1
+
+    # set source weights manually so the C++ that checks the
+    # size of the array is executed
+    vol_norm = False
+    strengths = np.zeros(n_cells*TETS_PER_VOXEL)
+    # set non-zero strengths only for the tets corresponding to the
+    # first two geometric hex cells
+    strengths[0:TETS_PER_VOXEL] = 10
```
Comment (paulromano) [inline tests/regression_tests/unstructured_mesh/source_sampling/test.py:63]:
```suggestion
    strengths[:TETS_PER_VOXEL] = 10

```

File: tests/regression_tests/unstructured_mesh/source_sampling/test.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+# This test uses a geometry file with cells that match a regular mesh. Each cell
+# in the geometry corresponds to 12 tetrahedra in the unstructured mesh file.
+@pytest.fixture
+def model():
+    openmc.reset_auto_ids()
+
+    ### Materials ###
+    materials = openmc.Materials()
+
+    water_mat = openmc.Material(material_id=3, name="water")
```
Comment (paulromano) [inline tests/regression_tests/unstructured_mesh/source_sampling/test.py:24]: Is it really necessary to assign an ID manually?

File: tests/regression_tests/unstructured_mesh/source_sampling/test.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    settings.source = source
+    return openmc.model.Model(geometry=geometry,
+                              materials=materials,
+                              settings=settings)
+
+
+test_cases = []
+for i, lib, in enumerate(['libmesh', 'moab']):
+    test_cases.append({'library' : lib,
+                       'inputs_true' : 'inputs_true{}.dat'.format(i)})
```
Comment (paulromano) [inline tests/regression_tests/unstructured_mesh/source_sampling/test.py:80]:
```suggestion
test_cases = [
    {'library': lib, 'inputs_true': f'inputs_true{i}.dat'}
    for i, lib in enumerate(('libmesh', 'moab'))
]

```

File: tests/unit_tests/test_source_mesh.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+    materials.append(water_mat)
+
+    ### Geometry ###
+    # This test uses a geometry file that resembles a regular mesh.
+    # 12 tets are used to match each voxel in the geometry.
+
+    # create a regular mesh that matches the superimposed mesh
+    regular_mesh = openmc.RegularMesh(mesh_id=10)
+    regular_mesh.lower_left = (-10, -10, -10)
+    regular_mesh.dimension = (10, 10, 10)
```
Comment (paulromano) [inline tests/unit_tests/test_source_mesh.py:39]: Same comments apply to this file as what appears in the corresponding regression test

File: tests/unit_tests/test_source_mesh.py [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+        if test_cases['source_strengths'] == 'uniform':
+            exp_vals = np.ones(n_cells) / n_cells
+        else:
+            # sum up the source strengths for each tet, these are the expected true mean
+            # of the sampling frequency for that cell
+            exp_vals = strengths.reshape(-1, 12).sum(axis=1) / sum(strengths)
+
+        diff = np.abs(mean - exp_vals)
+        assert((diff < 2*std_dev).sum() / diff[:10].size >= 0.95)
+        assert((diff < 6*std_dev).sum() / diff.size >= 0.97)
```
Comment (paulromano) [inline tests/unit_tests/test_source_mesh.py:147]: Neither of these seem to make sense to me -- can you explain?

**Review by paulromano — CHANGES_REQUESTED** [review]
Really liking this! I don't see any major flaws. I've suggested some changes throughout but I think we should be able to get this merged in soon.

File: include/openmc/distribution_spatial.h [hunk 11]
```diff
@@ -94,6 +95,32 @@ class SphericalIndependent : public SpatialDistribution {
...
+class MeshSpatial : public SpatialDistribution {
+public:
+  explicit MeshSpatial(pugi::xml_node node);
+
+  //! Sample a position from the distribution
+  //! \param seed Pseudorandom number seed pointer
+  //! \return Sampled position
+  Position sample(uint64_t* seed) const;
+
+  const unique_ptr<Mesh>& mesh() const { return model::meshes.at(mesh_idx_); }
```
[the remark this answers is not in the document]
Comment (pshriwise) [reply include/openmc/distribution_spatial.h:111]: Doesn't returning a unique pointer reference reduce the likelihood that the pointer would be accessed after the object is destroyed? References seem less likely to be used to instantiate other objects and such and the type implies that the unique pointer is managed elsewhere. Returning a `const unique_ptr<>&` seems to generally provide more information about the ownership of the object to me.

In this case, I think the chances someone would misuse or misunderstand the `const *` is small so either woudl do. But in a more general sense, what do you think?

File: include/openmc/distribution_spatial.h [hunk 12]
```diff
@@ -94,6 +95,32 @@ class SphericalIndependent : public SpatialDistribution {
...
+
+  int32_t n_sources() const { return this->mesh()->n_bins(); }
+
+private:
+  Mesh* mesh_ptr_ {nullptr};
+  int32_t mesh_idx_ {C_NONE};
```
[the remark this answers is not in the document]
Comment (pshriwise) [reply include/openmc/distribution_spatial.h:117]: For sure. Good call. Gonna stick with the index to be consistent with other classes.

File: include/openmc/mesh.h [hunk 13]
```diff
@@ -75,6 +75,18 @@ class Mesh {
 
   // Methods
 
+  //! Sample a mesh volume using a certain seed
+  //
+  //! \param[in] seed Seed to use for random sampling
+  //! \param[out] r Position within tet
+  virtual Position sample(uint64_t* seed, int32_t bin) const=0;
```
[the remark this answers is not in the document]
Comment (pshriwise) [reply include/openmc/mesh.h:82]: yep yep. On it. Thanks!

File: src/distribution_spatial.cpp [hunk 14]
```diff
@@ -180,6 +182,71 @@ Position SphericalIndependent::sample(uint64_t* seed) const
...
+{
+  // No in-tet distributions implemented, could include distributions for the barycentric coords
+  // Read in unstructured mesh from mesh_id value
+  int32_t mesh_id = std::stoi(get_node_value(node, "mesh_id"));
+  // Get pointer to spatial distribution
+  mesh_idx_ = model::mesh_map.at(mesh_id);
+
+  // Check whether mesh pointer points to a mesh
+  mesh_ptr_ = dynamic_cast<Mesh*>(model::meshes[mesh_idx_].get());
+  if (!mesh_ptr_) {fatal_error("Mesh passed to spatial distribution is not a mesh object"); }
```
[the remark this answers is not in the document]
Comment (pshriwise) [reply src/distribution_spatial.cpp:199]: We shouldn't. Gonna pull these lines. Thanks!

File: src/mesh.cpp [hunk 15]
```diff
@@ -2049,6 +2090,34 @@ std::string MOABMesh::library() const
   return mesh_lib_type;
 }
 
+// Sample position within a tet for MOAB type tets
+Position MOABMesh::sample(uint64_t* seed, int32_t bin) const {
+
+  moab::EntityHandle tet_ent = get_ent_handle_from_bin(bin);
+
+  // Get vertex coordinates for MOAB tet
+  vector<moab::EntityHandle> conn1;
```
[the remark this answers is not in the document]
Comment (pshriwise) [reply src/mesh.cpp:2099]: Yeah I think using a `std::array` here should be safe. Updating to the C-style call in MOAB and doing a size check for the size of the element 
connectivity.

File: src/mesh.cpp [hunk 16]
```diff
@@ -2501,6 +2570,21 @@ void LibMesh::initialize()
   bbox_ = libMesh::MeshTools::create_bounding_box(*m_);
 }
 
+// Sample position within a tet for LibMesh type tets
+Position LibMesh::sample(uint64_t* seed, int32_t bin) const {
+  const auto& elem = get_element_from_bin(bin);
+  // Get tet vertex coordinates from LibMesh
+  std::array<Position, 4> tet_verts;
+  for (int i = 0; i < elem.n_nodes(); i++) {
```
[the remark this answers is not in the document]
Comment (pshriwise) [reply src/mesh.cpp:2578]: For a linear tet, yeah. I'll add a check here that this is that case.

File: include/openmc/distribution_spatial.h [hunk 17]
```diff
@@ -94,6 +95,32 @@ class SphericalIndependent : public SpatialDistribution {
...
+  std::string sample_scheme_;
+  double total_strength_ {0.0};
+  std::vector<double> mesh_CDF_;
+  std::vector<double> mesh_strengths_;
```
[the remark this answers is not in the document]
Comment (pshriwise) [reply include/openmc/distribution_spatial.h:119]: Good point! A `Discrete` object would save some duplicated lines here. What do you recommend we do about the set of outcomes usually provided to `Discrete`?
 There would be some redundant storage of what is effectively a set of mesh bins.

File: openmc/stats/multivariate.py [hunk 18]
```diff
@@ -616,6 +619,131 @@ def from_xml_element(cls, elem):
...
+        -------
+        openmc.stats.MeshSpatial
+            Spatial distribution generated from XML element
+
+        """
+
+        mesh_id = int(elem.get('mesh_id'))
+
+        # check if this mesh has been read in from another location already
+        if mesh_id not in MESHES:
```
[the remark this answers is not in the document]
Comment (pshriwise) [reply openmc/stats/multivariate.py:735]: Yeah I was monkey-ing around with another approach and accidentally left that in. Thanks!

File: tests/regression_tests/unstructured_mesh/source_sampling/test.py (hunk 3 above)
Comment (pshriwise) [reply tests/regression_tests/unstructured_mesh/source_sampling/test.py:34]: That's a little tricky to do unless we change the unstructured meshes associated with these tests (i.e. `test_mesh_tets.e`). I could add a smaller mesh, but we already have these in the repo.

File: tests/regression_tests/unstructured_mesh/source_sampling/test.py (hunk 5 above)
Comment (pshriwise) [reply tests/regression_tests/unstructured_mesh/source_sampling/test.py:43]: Nice pointer. Thanks!

File: tests/regression_tests/unstructured_mesh/source_sampling/test.py (hunk 7 above)
Comment (pshriwise) [reply tests/regression_tests/unstructured_mesh/source_sampling/test.py:24]: Nope. Removing...

**Comment by pshriwise** [comment]
> @pshriwise One thing I noticed right away is that the formatting of some of the C++ source files seems off. Do you have the clang-format [post-commit hook installed](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting)?

Sorry about that. I do have the commit hook installed. Though I've added lines to disable it during a rebase so it doesn't mess with the way I've resolved conflicts, so maybe that got me here. I just ran it on the set of source files changed in this PR so it should be aright now.

File: include/openmc/distribution_spatial.h (hunk 17 above)
[the remark this answers is not in the document]
Comment (paulromano) [reply include/openmc/distribution_spatial.h:119]: Hmm, I see what you mean. We would need to pass to `Discrete` a vector of increasing integer indices, which seems a little silly. I'd say just go with that for now and in the future, we might consider breaking out a separate class from `Discrete` that only contains N probabilities and returns integers in the interval [0, N-1].

File: include/openmc/distribution_spatial.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (paulromano) [reply include/openmc/distribution_spatial.h:111]: Returning a `const unique_ptr<T>&` to me is no different to me than `const T*` or `const T&`. In either case, the caller does not own the object and would need to be careful not to use the object past its lifetime. Making it `unique_ptr` just makes it a little more cumbersome to work with for the caller.

File: tests/regression_tests/unstructured_mesh/source_sampling/test.py (hunk 3 above)
Comment (paulromano) [reply tests/regression_tests/unstructured_mesh/source_sampling/test.py:34]: Ah yeah, I see. Ok, no worries!

File: tests/regression_tests/unstructured_mesh/source_sampling/test.py [hunk 19]
```diff
@@ -0,0 +1,114 @@
+from itertools import product
```
[the remark this answers is not in the document]
Comment (pshriwise) [reply tests/regression_tests/unstructured_mesh/source_sampling/test.py:1]: Yeah, let me think about something meaningful to tally here -- probably something with a `CellBorn` filter.

File: include/openmc/distribution_spatial.h (hunk 17 above)
[the remark this answers is not in the document]
Comment (pshriwise) [reply include/openmc/distribution_spatial.h:119]: Yeah, I was thinking that too about the separate class. We could also return an index from Discrete if the outcomes vector is empty? But maybe that's a little confusing.
Okay, I'll make a note of that and we'll move on with what we have for now.

File: tests/regression_tests/unstructured_mesh/source_sampling/test.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (pshriwise) [reply tests/regression_tests/unstructured_mesh/source_sampling/test.py:1]: I ended up adding a flux tally here. It isn't that meaningful for this model though. It's main purpose is to ensure the `openmc` executable will run successfully with a mesh source -- the unit tests only sample source sites.

Maybe it would be better to do both the source site sampling and a full run as part of the unit tests instead?

File: include/openmc/distribution_spatial.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (pshriwise) [reply include/openmc/distribution_spatial.h:111]: True, a `const T*` shouldn't be interpreted as ownership really. I'll update to a `const T*`.

File: src/distribution_spatial.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (NybergWISC) [reply src/distribution_spatial.cpp]: I believe this line is left over from when we used it to check whether the mesh pointer was of type unstructured mesh. Since we added functionality/error messages for all types of meshes it became redundant.

File: tests/regression_tests/unstructured_mesh/source_sampling/test.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (pshriwise) [reply tests/regression_tests/unstructured_mesh/source_sampling/test.py]: I went a head and added a call to run particles in the `test_mesh_source.py` to check that transport is successful for a handful of particles. The check for an incorrect strengths array size was moved there as well and I've removed this regression test. I think that makes the most sense for this capability.

File: tests/unit_tests/test_source_mesh.py (hunk 10 above)
Comment (pshriwise) [reply tests/unit_tests/test_source_mesh.py]: We talked about these offline, but for posterity:
   - The [:10] was leftover from debugging an issue. My bad
   - The `0.97` was a typo and should be `0.997`.

File: tests/unit_tests/test_source_mesh.py [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+from pathlib import Path
+import pytest
+import numpy as np
+
+import openmc
+import openmc.lib
+
+from tests import cdtemp
+from tests.regression_tests import config
+from subprocess import call
```
Comment (paulromano) [inline tests/unit_tests/test_source_mesh.py:12]:
```suggestion
from itertools import product
from pathlib import Path
from subprocess import call

import pytest
import numpy as np
import openmc
import openmc.lib

from tests import cdtemp
from tests.regression_tests import config
```

File: tests/unit_tests/test_source_mesh.py [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+    with cdtemp([mesh_filename]):
+        model.export_to_xml()
+
+        n_measurements = 100
+        n_samples = 1000
+
+        cell_counts = np.zeros((n_cells, n_measurements))
+
+        # This model contains 1000 geometry cells. Each cell is a hex
+        # corresponding to 12 of the tets. This test runs 10000 particles. This
```
Comment (paulromano) [inline tests/unit_tests/test_source_mesh.py:110]: 10000 particles? Should it be 1000 samples?

**Review by paulromano — CHANGES_REQUESTED** [review]
Almost there! A few more comments:

File: tests/unit_tests/test_source_mesh.py (hunk 21 above)
Comment (pshriwise) [reply tests/unit_tests/test_source_mesh.py:109]: Mmm yep that's an outdated comment. Thanks!

File: openmc/stats/multivariate.py [hunk 22]
```diff
@@ -617,6 +620,130 @@ def from_xml_element(cls, elem):
...
```
[the remark this answers is not in the document]
Comment (pshriwise) [reply openmc/stats/multivariate.py:735]: Agreed. Adding this test revealed another issue reading in element source strengths as well. 👍🏻

File: openmc/mesh.py [hunk 23]
```diff
@@ -1937,3 +1938,25 @@ def from_xml_element(cls, elem):
         length_multiplier = float(get_text(elem, 'length_multiplier', 1.0))
 
         return cls(filename, library, mesh_id, '', length_multiplier)
+
+
+def read_meshes(elem):
```
Comment (paulromano) [inline openmc/mesh.py:1943]: Since this is only used in settings.py, why not just move it there? Also, since it is really not intended to used by a user, please change the name to `_read_meshes` if it's going to live at module scope.

File: openmc/mesh.py [hunk 24]
```diff
@@ -1937,3 +1938,25 @@ def from_xml_element(cls, elem):
...
+    elem : xml.etree.ElementTree.Element
+        XML element
+
+    Returns
+    -------
+    OrderedDict
+        An ordered dictionary with mesh IDs as keys and openmc.MeshBase
+        instanaces as values
+    """
+    out = OrderedDict()
```
Comment (paulromano) [inline openmc/mesh.py:1957]: Does it really need to be ordered?

**Review by paulromano — CHANGES_REQUESTED** [review]
One last round and then we really should be good to merge

File: openmc/mesh.py (hunk 23 above)
Comment (pshriwise) [reply openmc/mesh.py:1943]: I'm thinking that there are other places we might be able to use this as well going forward and because it pertains only to meshes I think it makes sense to keep it in this module. That'd be my vote anyways. Feel free to override 😄 

I do see this as a utility function for I/O operations within our classes, so I'll go ahead and make that name change. Thanks for catching that!

File: openmc/mesh.py (hunk 24 above)
Comment (pshriwise) [reply openmc/mesh.py:1957]: Mmmm nope. Updating...

**Review by paulromano — APPROVED** [review]
Thanks again @pshriwise and @NybergWISC!

File: src/distribution_spatial.cpp [hunk 25]
```diff
@@ -180,6 +183,84 @@ Position SphericalIndependent::sample(uint64_t* seed) const
...
+      mesh_strengths_[i] *= mesh()->volume(i);
+    }
+  }
+
+  total_strength_ =
+    std::accumulate(mesh_strengths_.begin(), mesh_strengths_.end(), 0.0);
+
+  for (int i = 0; i < n_bins; i++) {
+    mesh_CDF_[i + 1] = mesh_CDF_[i] + mesh_strengths_[i] / total_strength_;
+  }
```
Comment (gonuke) [inline src/distribution_spatial.cpp:245]: This is prone to round-off error - issue and PR incoming!

Comment (pshriwise) [reply src/distribution_spatial.cpp:245]: I think I noted this in the original PR and forgot to fix it in mine 🤦🏻. Looking forward to the fix! Thank you!
